File: digidis-go/day9/day9.go
package main

import (
	"fmt"
	"io/ioutil"
)

func main() {
	var level, score, garbage int
	row, _ := ioutil.ReadFile("input.txt")

	for i := 0; i < len(row); i++ {
		switch row[i] {
		case '{':
			level++

		case '}':
			score += level
			level--

		case '<':
			i++
			for row[i] != '>' {
				if row[i] == '!' {
					i++
				} else {
					garbage++
				}
				i++
			}

		case '!':
			i++
		}
	}

	fmt.Printf("score: %v\n", score)
	fmt.Printf("garbage: %v\n", garbage)
}

File: dtrng-javascript/Day9.js
const solveA = ({data}) => {
  return data
    .split('')
    .reduce((acc, char) => {
      if (acc.ignore) {
        acc.ignore = false
      } else if (acc.inGarbage) {
        if (char === '>') {
          acc.inGarbage = false
        } else if (char === '!') {
          acc.ignore = true
        }
      } else {
        if (char === '<') {
          acc.inGarbage = true
        } else if (char === '{') {
          acc.groupLevel++
        } else if (char === '}') {
          acc.score += acc.groupLevel
          acc.groupLevel--
        }
      }
      return acc
    }, {groupLevel: 0, inGarbage: false, ignore: false, score: 0})
    .score
}

const solveB = ({data}) => {
  return data
    .split('')
    .reduce((acc, char) => {
      if (acc.ignore) {
        acc.ignore = false
      } else if (acc.inGarbage) {
        if (char === '>') {
          acc.inGarbage = false
        } else if (char === '!') {
          acc.ignore = true
        } else {
          acc.removedChars++
        }
      } else if (char === '<') {
        acc.inGarbage = true
      }
      return acc
    }, {groupLevel: 0, inGarbage: false, ignore: false, removedChars: 0})
    .removedChars
}

require('fs').readFile('input.txt', 'utf8', (err, data) => {
  if (err) { console.error(err); return }
  data = data.split('\n')[0]
  console.log('solution to a:', solveA({data}))
  console.log('solution to b:', solveB({data}))
})
